### PR TITLE
Introduce consistent .rename_axes and .rename API for maps

### PR DESCRIPTION
**Pull request by QRemy**
Rename Geom axes
-could be used in https://github.com/gammapy/gammapy/pull/4018#discussion_r913135787
-could be used in `Geom.as_energy_true`
-could  also be useful to rename axes of FITS maps that do not follow the expected convention.

### CODE REVIEW

**Comment by adonath** [comment]
Thanks @QRemy! I remember we discussed this functionality before:
- https://github.com/gammapy/gammapy/issues/3781
- https://github.com/gammapy/gammapy/pull/3799

The conclusion was the recommended pattern to rename axis should be `new_axis = axis.copy(name="new-name")`. So making an explicit copy. As we rely on axis name conventions in Gammapy the in-place name change could break things unexpectedly. I think the renaming is typically done from existing objects, such as:

```python3
energy_axis = dataset.psf.psf_map.geom.axes["energy_true"]
energy_axis.rename("energy")

new_map = WcsNDMap.create(axes=[energy_axis])

# This fails then for not so obvious reasons...
dataset.npred()
```
So we decided the `.copy()` pattern should be preferred:

```python3
energy_axis = dataset.psf.psf_map.geom.axes["energy_true"]
energy_axis = energy_axis.copy(name="energy")

new_map = WcsNDMap.create(axes=[energy_axis])

# Does not fail...
dataset.npred()
```

Which advantage do you see in the in-place name update? Or do we just need to document the `.copy(name="new-name")` pattern better?

**Comment by QRemy** [comment]
If we want to modify the axes name attached to a Map or a Geom it is not trivial (just looking how `Geom.as_energy_true` is implement). `Geom.rename_axes` is more generic and closer to the pattern in astropy `Table.rename_columns`. I could add an option `copy=True` by default to match the current behavior.  The in place change could be usefull to change the axes names of Map that do not follows the gammapy conventions without coying them.

**Comment by adonath** [comment]
@QRemy I think there are two separate things to be discussed:
- In-place modification vs copy
- Convenience API for renaming axes / axes in existing geometries

I fully agree that convenience methods such as:
```python

new_axes = axes.rename_axes(names={"energy": "energy_true", "foo": "bar"})
new_geom = geom.rename_axes(names={"energy": "energy_true", "foo": "bar"})
new_map = map.rename_axes(names={"energy": "energy_true", "foo": "bar"})
```
...would be very useful!

I think a dedicated `axis.rename(name="new-name")`, would not be needed, because it will be an alias for `axis.copy(name="new-name")`, but could be introduced for API consistency with the `.rename_axes()`.

However we should still (as shown in the example above) keep the copy and return behavior. This is just how our API is defined now. All methods on `MapAxis`, `Geom` and `Map` return a new modified object and do not support in place operations. There is exactly one exception, which is `Map.stack()`. The only reason it exists is performance. Always creating a full copy of the data would be the bottleneck and slow down stacking a lot. 

It is true that e.g. `Table.rename_column()` does an in-place update, but e.g. `Table.group_by()` does not. I'm not sure whether they tried to achieve API consistency here at all.

File: gammapy/maps/geom.py [hunk 1]
```diff
@@ -349,11 +349,31 @@ def slice_by_idx(self, slices):
...
+            New names of the axes (list must be of same length than `names`).
+
+        Returns
+        -------
+        geom : `~Geom`
+            Renamed geometry.
+        """
+        geom = self.copy()
+        geom._axes = geom.axes.rename(names, new_names)
+        return geom
```
Comment (adonath) [inline gammapy/maps/geom.py:370]:
```suggestion
        axes = self.axes.rename(names=names, new_names=new_names)
        return self._init_copy(axes=axes)
```

File: gammapy/maps/geom.py [hunk 2]
```diff
@@ -349,11 +349,31 @@ def slice_by_idx(self, slices):
...
+
     @property
     def as_energy_true(self):
         """If the geom contains an energy axis rename it to energy true"""
-        energy_axis = self.axes["energy"].copy(name="energy_true")
-        return self.to_image().to_cube([energy_axis])
+        return  self.rename_axes("energy", "energy_true", copy=True)
```
Comment (adonath) [inline gammapy/maps/geom.py:375]:
```suggestion
        return  self.rename_axes(name="energy", new_name="energy_true")
```

File: gammapy/maps/core.py [hunk 3]
```diff
@@ -132,6 +132,25 @@ def quantity(self, val):
...
+            New names of the axes (list must be of same length than `names`).
+
+        Returns
+        -------
+        geom : `~Map`
+            Renamed Map.
+        """
+        map_copy = self.copy()
+        map_copy._geom = map_copy.geom.rename_axes(names, new_names)
+        return map_copy
```
Comment (adonath) [inline gammapy/maps/core.py:152]:
```suggestion
        geom = self.geom.rename_axes(names=names, new_names=new_names)
        return self._init_copy(geom=geom)
```

File: gammapy/maps/axes.py [hunk 4]
```diff
@@ -2044,6 +2059,31 @@ def assert_names(self, required_names):
...
+        axes : `MapAxes`
+            Renamed Map axes object
+        """
+        axes = self.copy()
+        if isinstance(names, str):
+            names = [names]
+        if isinstance(new_names, str):
+            new_names = [new_names]
+        for name, new_name in zip(names, new_names):
+            axes[name]._name = new_name
```
Comment (adonath) [inline gammapy/maps/axes.py:2084]: This does not really make a copy of the `MapAxis`, just use the `MapAxis.rename()` method here you introduced earlier.

File: gammapy/maps/axes.py [hunk 5]
```diff
@@ -339,6 +339,21 @@ def to_node_type(self, node_type):
...
+        Parameters
+        ----------
+        new_name : str
+            The new name for the axis.
+
+        Returns
+        -------
+        axis : `~gammapy.maps.MapAxis`
+            Renamed MapAxis
+        """
```
Comment (adonath) [inline gammapy/maps/axes.py:355]: We could maybe add a check to `MapAxis.__init__`, that `name` is actually as string...not sure we have this.

**Review by adonath — CHANGES_REQUESTED** [review]
Thanks @QRemy, I have left some inline comments and suggestions for simplification (please accept those if you agree...).

File: gammapy/maps/axes.py (hunk 4 above)
Comment (QRemy) [reply gammapy/maps/axes.py:2084]: but here axes is self.copy() which already copied each axis

File: gammapy/maps/geom.py [hunk 6]
```diff
@@ -349,11 +349,30 @@ def slice_by_idx(self, slices):
         axes = self.axes.slice_by_idx(slices)
         return self._init_copy(axes=axes)
 
+
+    def rename_axes(self, names, new_names):
+        """Rename the axes in place.
```
Comment (adonath) [inline gammapy/maps/geom.py:354]:
```suggestion
        """Rename axes contained in the geometry
```

**Comment by QRemy** [comment]
Thanks @adonath for all the corrections

**Comment by adonath** [comment]
@registerrier Can you please do the final review here? The PR is fine from my side now...

File: gammapy/maps/core.py [hunk 7]
```diff
@@ -132,6 +132,24 @@ def quantity(self, val):
         self.data = val.value
         self._unit = val.unit
 
+    def rename_axes(self, names, new_names):
+        """Rename the Map axes.
+
```
Comment (registerrier) [inline gammapy/maps/core.py:137]: Maybe specify that this will create a full copy of the `Map` as this might not be obvious from the name.

**Review by registerrier — APPROVED** [review]
Thanks @QRemy . This looks good to me.

I have just left a small inline comment.

File: gammapy/maps/core.py (hunk 7 above)
Comment (adonath) [reply gammapy/maps/core.py:137]: By default all maps methods do a copy, except for `.stack()`,  so I think it's fine to just state the exception.

**Review by adonath — APPROVED** [review]

**Pull request merged**